### PR TITLE
Add setting to configure the max memory for tsserver

### LINKED ISSUE #33524: Pass max_old_space_size to TSServer
Per discussion here: https://github.com/Microsoft/TypeScript/issues/18055

Some things I noticed while looking into this:
* I believe either `--max_old_space_size` or `--max-old-space-size` works, v8 args can be dashes or underscores. Underscores work for sure.
* TSServer appears to have special logic if `global.gc` is available, so maybe consider adding `--expose-gc` as well.
* I believe the maximum settings for 64-bit builds of Node is `4096`
* There's this: https://github.com/electron/electron/issues/9693. Essentially when running VS Code with the `./script/code.sh` script, setting these flags in `execArgv` options in `typescriptserviceclient.ts` had no effect. ~I think it'll work if Electron is packaged and ran normally, but I'm not positive.~

### PR DESCRIPTION
**Pull request by MLoughry**
This PR fixes #33524 by adding a setting `typescript.tsserver.maxTsServerMemory`, which passes the setting on to `tsserver` using `--max-old-space-size`

### CODE REVIEW

File: extensions/typescript-language-features/package.json [hunk 1]
```diff
@@ -685,6 +685,12 @@
           "default": true,
           "description": "%configuration.tsserver.useSeparateSyntaxServer%",
           "scope": "window"
+        },
```
Comment (amcasey) [inline extensions/typescript-language-features/package.json]: FWIW, in VS we didn't expose this as an option - we just increased it.  It's a cap and not an amount to allocate, so there was (as far as we could tell) no cost to increasing it without using it.

File: extensions/typescript-language-features/package.nls.json [hunk 2]
```diff
@@ -54,6 +54,7 @@
 	"typescript.problemMatchers.tscWatch.label": "TypeScript problems (watch mode)",
 	"configuration.suggest.paths": "Enable/disable suggestions for paths in import statements and require calls.",
 	"configuration.tsserver.useSeparateSyntaxServer": "Enable/disable spawning a separate TypeScript server that can more quickly respond to syntax related operations, such as calculating folding or computing document symbols. Requires using TypeScript 3.4.0 or newer in the workspace.",
+	"configuration.tsserver.maxTsServerMemory": "Set the maximum amount of memory to allocate to the TypeScript server process",
```
Comment (amcasey) [inline extensions/typescript-language-features/package.nls.json]: Consider including the units (MB, I assume).

File: extensions/typescript-language-features/src/tsServer/spawner.ts [hunk 3]
```diff
@@ -85,10 +85,13 @@ export class TypeScriptServerSpawner {
...
 	}
 
-	private getForkOptions(kind: ServerKind) {
+	private getForkOptions(kind: ServerKind, configuration: TypeScriptServiceConfiguration) {
 		const debugPort = TypeScriptServerSpawner.getDebugPort(kind);
 		const tsServerForkOptions: electron.ForkOptions = {
-			execArgv: debugPort ? [`--inspect=${debugPort}`] : [],
+			execArgv: [
+				...(debugPort ? [`--inspect=${debugPort}`] : []),
+				...(configuration.maxTsServerMemory ? [`--max-old-space-size=${configuration.maxTsServerMemory}`] : [])
```
Comment (amcasey) [inline extensions/typescript-language-features/src/tsServer/spawner.ts]: I think this has already validated to be an int (or, at least, a number), but I thought I should confirm.  We have some reason to believe we're not putting a raw string into the command line?

**Review by amcasey — COMMENTED** [review]
Minor comments.

**Comment by mjbvz** [comment]
Thanks @MLoughry! Just want to confirm, using a high memory setting does fix the out of memory issues for you?

File: extensions/typescript-language-features/package.json (hunk 1 above)
Comment (MLoughry) [reply extensions/typescript-language-features/package.json]: That's not entirely true, as far as I know from my experience. My understanding is that Node will only trigger a GC if it hits this limit, so a process that cycles through memory will eventually hit this limit.

**Comment by MLoughry** [comment]
@mjbvz Yes. `tsserver` still takes a while to spin up on our massive project, but it no longer crashes and "forgets" all the work it did when I configure a higher memory limit.

File: extensions/typescript-language-features/src/tsServer/spawner.ts (hunk 3 above)
Comment (MLoughry) [reply extensions/typescript-language-features/src/tsServer/spawner.ts]: I added additional validation to `readMaxTsServerMemory` to ensure the value in configuration is a non-negative safe integer.

File: extensions/typescript-language-features/src/utils/configuration.ts [hunk 4]
```diff
@@ -146,4 +149,12 @@ export class TypeScriptServiceConfiguration {
 	private static readUseSeparateSyntaxServer(configuration: vscode.WorkspaceConfiguration): boolean {
 		return configuration.get<boolean>('typescript.tsserver.useSeparateSyntaxServer', true);
 	}
+
+	private static readMaxTsServerMemory(configuration: vscode.WorkspaceConfiguration): number {
+		const memoryInMB = configuration.get<number>('typescript.tsserver.maxTsServerMemory', 0);
+		if (!Number.isSafeInteger(memoryInMB) || memoryInMB < 0) {
+			return 0;
```
Comment (mjbvz) [inline extensions/typescript-language-features/src/utils/configuration.ts]: We probably should add a lower bound here, such as 128mb. We do this for a few other settings like fontSize

**Comment by mjbvz** [comment]
Thanks!

**Pull request merged**